### ISSUE TITLE
Converting a model with Int8 fake quant nodes from a saved model file fails

### ISSUE THREAD

**Issue by AdamHillier** [tensorflow/tensorflow#46707]
**System information**

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOS 10.15.6
- TensorFlow installed from (source or binary): TensorFlow 2.4.1 PyPI pip package.
- TensorFlow version (or github SHA if from source): 2.4.1.

**Command used to run the converter or code if you’re using the Python API**

Suppose `keras_model` is a Keras model containing QAT fake quantiser ops, like `tf.quantization.fake_quant_with_min_max_vars`. Then, the following conversion code using `from_saved_model` fails to output a valid Int8 model and throws an error.

```python
with open("converted_saved_model.tflite", "wb") as f:
    tf.keras.models.save_model(keras_model, "tmp-saved-model", save_format="tf")
    converter = tf.lite.TFLiteConverter.from_saved_model("tmp-saved-model")
    converter.optimizations = [tf.lite.Optimize.DEFAULT]
    converter.inference_input_type = tf.int8
    converter.inference_output_type = tf.int8
    f.write(converter.convert())
```

Note that `from_keras_model` works correctly.

See a minimal reproduction with the following colab notebook: https://colab.research.google.com/drive/1s3AkxetmaIHKcB3M69cBSVbOfhNx8106?usp=sharing

**The output from the converter invocation**

The above snippet fails with:

```
ValueError: The inference_input_type and inference_output_type must be tf.float32.
```

For more detail, see the linked colab notebook.

**Also, please include a link to the saved model or GraphDef**

Can be found in the linked colab notebook.

**Failure details**

Attempting to convert and setting the inference input/output type to int8 causes the error above.

Attempting to convert without setting those types (such that they default to float) doesn't cause an error but does yield a broken Int8 model with lots of dangling quant/dequant ops pairs:

<img width="500" alt="image" src="https://user-images.githubusercontent.com/7688302/105897373-f9673a80-600f-11eb-95ad-683671cbf825.png">

**Comment by AdamHillier** [comment]
This seems to stem from the fact that the following check does not return true:

https://github.com/tensorflow/tensorflow/blob/8027470e1e43a17645bf843815a236b92390d311/tensorflow/lite/python/lite.py#L410-L420

When converting from the saved model, the `node_def.op` values are the following:

```
VarHandleOp
ReadVariableOp
VarHandleOp
ReadVariableOp
VarHandleOp
ReadVariableOp
VarHandleOp
ReadVariableOp
VarHandleOp
ReadVariableOp
VarHandleOp
ReadVariableOp
VarHandleOp
ReadVariableOp
NoOp
Const
Placeholder
StatefulPartitionedCall
Placeholder
StatefulPartitionedCall
StatefulPartitionedCall
```

whereas when using `from_keras_model` they are:

```
Placeholder
Const
Const
Const
Const
Const
Const
Const
Identity
Identity
Identity
Identity
Identity
Identity
Identity
FakeQuantWithMinMaxVars
```

(using the model from the colab notebook)

This is why the check is true when running `from_keras_model`, but false when running `from_saved_model`.

Slightly related: #46704.

**Comment by ravikyram** [comment]
I have tried in colab with TF version 2.4, Nightly version(`2.5.0-dev20210126`) and was able to reproduce the issue. Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/1f48dcd87a3bc22d0fbf5363f8ef2b32/untitled631.ipynb). Thanks!

**Comment by abattery** [comment]
@MeghnaNatraj could you take a look at this?

**Comment by MeghnaNatraj** [comment]
It looks like the model has not been generated correctly. Reference [colab](https://colab.research.google.com/gist/MeghnaNatraj/6245ffa33364f2e2c7147f37c7b243c2/untitled631.ipynb).

Refer to the last cell in this notebook. 


```
with open("converted_saved_model.tflite", "wb") as f:
    tf.saved_model.save(keras_model, "tmp-saved-model")
    keras_model_ = tf.saved_model.load("tmp-saved-model")
    converter = tf.lite.TFLiteConverter.from_keras_model(keras_model_)
    # tf.keras.models.save_model(keras_model, "tmp-saved-model")
    # converter = tf.lite.TFLiteConverter.from_saved_model("tmp-saved-model")
    converter.optimizations = [tf.lite.Optimize.DEFAULT]
    converter.inference_input_type = tf.int8
    converter.inference_output_type = tf.int8
    f.write(converter.convert())
```

```
WARNING:absl:Found untraced functions such as int8_fake_quantize_1_layer_call_fn, int8_fake_quantize_1_layer_call_and_return_conditional_losses, int8_fake_quantize_1_layer_call_fn, int8_fake_quantize_1_layer_call_and_return_conditional_losses, int8_fake_quantize_1_layer_call_and_return_conditional_losses while saving (showing 5 of 5). These functions will not be directly callable after loading.
WARNING:absl:Found untraced functions such as int8_fake_quantize_1_layer_call_fn, int8_fake_quantize_1_layer_call_and_return_conditional_losses, int8_fake_quantize_1_layer_call_fn, int8_fake_quantize_1_layer_call_and_return_conditional_losses, int8_fake_quantize_1_layer_call_and_return_conditional_losses while saving (showing 5 of 5). These functions will not be directly callable after loading.
INFO:tensorflow:Assets written to: tmp-saved-model/assets
INFO:tensorflow:Assets written to: tmp-saved-model/assets
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-5-400338f2c78a> in <module>()
      8     converter.inference_input_type = tf.int8
      9     converter.inference_output_type = tf.int8
---> 10     f.write(converter.convert())

/usr/local/lib/python3.6/dist-packages/tensorflow/lite/python/lite.py in convert(self)
    837     # to None.
    838     # Once we have better support for dynamic shapes, we can remove this.
--> 839     if not isinstance(self._keras_model.call, _def_function.Function):
    840       # Pass `keep_original_batch_size=True` will ensure that we get an input
    841       # signature including the batch dimension specified by the user.

AttributeError: '_UserObject' object has no attribute 'call'
```

Will get back once I have an update on how this can be fixed.

**Comment by Xhark** [comment]
It just works if you use this save/load function.

    with open("converted_saved_model.tflite", "wb") as f:
        tf.keras.models.save_model(keras_model, "tmp-saved-model", save_format="tf")
        keras_model_ = tf.keras.models.load_model("tmp-saved-model")
        converter = tf.lite.TFLiteConverter.from_keras_model(keras_model_)
        
        converter.optimizations = [tf.lite.Optimize.DEFAULT]
        converter.inference_input_type = tf.int8
        converter.inference_output_type = tf.int8
        f.write(converter.convert())


I think there's a bug on converter validator that seems not check nested function.
I'll try to fix it. Thanks.

**Comment by Xhark** [comment]
Here's the fix: 6be1ebb8498e2ccdb5d05ad673678b98a9373346
Please reopen the bug if you have any problem with this changes.

**Comment by AdamHillier** [comment]
I tested this by pasting in your patch into my local `lite.py` and it fixed both issues I saw. Thanks for the help @MeghnaNatraj @Xhark!
